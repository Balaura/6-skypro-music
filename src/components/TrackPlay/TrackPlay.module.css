.trackPlay {
     display: flex;
     flex-direction: row;
   }
   
   .contain {
     width: auto;
     display: grid;
     grid-template-columns: auto 1fr;
     grid-template-areas: "image author" "image album";
     align-items: center;
   }
   
   .image {
     width: 51px;
     height: 51px;
     background-color: #313131;
     display: flex;
     align-items: center;
     justify-content: center;
     margin-right: 12px;
     grid-row: 1 / 3;
     grid-column: 1;
     grid-area: image;
   }
   
   .svg {
     width: 18px;
     height: 17px;
     fill: transparent;
     stroke: #4e4e4e;
   }
   
   .author {
     grid-row: 1;
     grid-column: 2;
     grid-area: author;
     min-width: 49px;
   }
   
   .authorLink {
     font-style: normal;
     font-weight: 400;
     font-size: 16px;
     line-height: 24px;
     color: #ffffff;
     white-space: nowrap;
   }
   
   .album {
     grid-row: 2;
     grid-column: 2;
     grid-area: album;
     min-width: 49px;
   }
   
   .albumLink {
     font-style: normal;
     font-weight: 400;
     font-size: 13px;
     line-height: 24px;
     color: #ffffff;
   }
   
  .likeDis {
    display: flex;
    flex-direction: row;
    align-items: center;
    position: relative;
    padding-left: 40px;
  }
  
  .like,
  .dislike {
    padding: 5px;
    position: relative;
    width: 14px;
    height: 12px;
    cursor: pointer;
    background-color: transparent;
    border: none;
    background: none;
  }
  
  .likeBtn {
    background: none;
    border: none;
    cursor: pointer;
    padding: 0;
    display: flex;
    align-items: center;
    justify-content: center;
  }
  
  .likeSvg {
    width: 14px;
    height: 12px;
    fill: transparent;
    stroke: #696969;
    stroke-width: 1px;
  }
  
  .likeBtn:hover .likeSvg {
    stroke: #ACACAC;
  }
  
  .likeBtn.liked .likeSvg {
    fill: #B672FF;
    stroke: #B672FF;
  }