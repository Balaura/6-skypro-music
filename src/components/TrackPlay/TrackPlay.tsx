import React, { useState } from 'react';
import styles from './TrackPlay.module.css';
import { Track } from '@/hooks/useFetchTracks';
import { useDispatch, useSelector } from 'react-redux';
import { RootState } from '@/store/store';
import { addTrackToFavorites, removeTrackFromFavorites } from '@/api/api';
import { addToFavorites, removeFromFavorites } from '@/store/features/audioPlayerSlice';

interface TrackPlayProps {
  currentTrack: Track;
}

const TrackPlay: React.FC<TrackPlayProps> = ({ currentTrack }) => {
  const _id = currentTrack._id;
  const dispatch = useDispatch();
  const favoriteTracks = useSelector((state: RootState) => state.audioPlayer.favoriteTracks);
  const isFavorite = favoriteTracks.includes(currentTrack._id);
  const [isLoading, setIsLoading] = useState(false);
  const [error, setError] = useState<string | null>(null);

  const handleLike = async (e: React.MouseEvent) => {
    e.stopPropagation();
    setIsLoading(true);
    setError(null);
    try {
      if (isFavorite) {
        await removeTrackFromFavorites(_id);
        dispatch(removeFromFavorites(_id));
      } else {
        await addTrackToFavorites(_id);
        dispatch(addToFavorites(_id));
      }
    } catch (error) {
      console.error('Ошибка при обновлении избранного:', error);
      setError('Не удалось обновить избранное');
    } finally {
      setIsLoading(false);
    }
  };

  return (
    <div className={styles.trackPlay}>
      <div className={styles.contain}>
        <div className={styles.image}>
          <svg className={styles.svg}>
            <use xlinkHref="img/icon/sprite.svg#icon-note"></use>
          </svg>
        </div>
        <div className={styles.author}>
          <a className={styles.authorLink} href="#">
            {currentTrack?.name}
          </a>
        </div>
        <div className={styles.album}>
          <a className={styles.albumLink} href="#">
            {currentTrack?.author}
          </a>
        </div>
      </div>
      <div className={styles.likeDis}>
        <button
          className={`${styles.likeBtn} ${isFavorite ? styles.liked : ''}`}
          onClick={handleLike}
          disabled={isLoading}
        >
          <svg className={styles.likeSvg}>
            <use xlinkHref={isFavorite ? "img/icon/sprite.svg#icon-like" : "img/icon/sprite.svg#icon-dislike"}></use>
          </svg>
        </button>
        <div className={`${styles.like} ${styles._btnIcon}`}>
          <svg className={styles.likeSvg}>
            <use xlinkHref="img/icon/sprite.svg#icon-like"></use>
          </svg>
        </div>
        <div className={`${styles.dislike} ${styles._btnIcon}`}>
          <svg className={styles.dislikeSvg}>
            <use xlinkHref="img/icon/sprite.svg#icon-dislike"></use>
          </svg>
        </div>
      </div>
    </div>
  );
};

export default TrackPlay;