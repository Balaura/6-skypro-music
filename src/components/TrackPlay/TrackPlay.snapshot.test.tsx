import React from 'react';
import renderer from 'react-test-renderer';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import TrackPlay from './TrackPlay';

const mockStore = configureStore([]);

describe('TrackPlay component', () => {
  it('renders correctly', () => {
    const store = mockStore({
      audioPlayer: {
        favoriteTracks: [],
      },
      auth: {
        username: 'testUser',
      },
    });

    const currentTrack = {
      _id: 1,
      name: 'Test Track',
      author: 'Test Author',
      release_date: '2023-01-01',
      genre: 'Test Genre',
      duration_in_seconds: 180,
      album: 'Test Album',
      track_file: 'http://example.com/test.mp3',
      artist: 'Test Artist',
    };

    const tree = renderer.create(
      <Provider store={store}>
        <TrackPlay currentTrack={currentTrack} />
      </Provider>
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });
});