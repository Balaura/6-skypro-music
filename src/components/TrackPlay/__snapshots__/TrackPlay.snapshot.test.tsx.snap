// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TrackPlay component renders correctly 1`] = `
<div
  className="trackPlay"
>
  <div
    className="contain"
  >
    <div
      className="image"
    >
      <svg
        className="svg"
      >
        <use
          xlinkHref="/img/icon/sprite.svg#icon-note"
        />
      </svg>
    </div>
    <div
      className="author"
    >
      <a
        className="authorLink"
        href="#"
      >
        Test Track
      </a>
    </div>
    <div
      className="album"
    >
      <a
        className="albumLink"
        href="#"
      >
        Test Author
      </a>
    </div>
  </div>
  <div
    className="likeDis"
  >
    <button
      className="likeBtn "
      disabled={false}
      onClick={[Function]}
    >
      <svg
        className="likeSvg"
      >
        <use
          xlinkHref="/img/icon/sprite.svg#icon-like"
        />
      </svg>
    </button>
  </div>
</div>
`;
